### PR TITLE
If a grammar cannot be found, fall back to plain text parser.

### PR DESCRIPTION
**Pull request by snim2**
The intention here is to enable the user to diff files for which we don't yet have a grammar.

### CODE REVIEW

**Comment by ltratt** [comment]
A dumb question, but just to check: does this definitely work on arbitrary text inputs?

**Comment by snim2** [comment]
I've been trying it out with Rust files and their `.bk` backups and it seems to work OK there.

File: src/lib/ast.rs [hunk 1]
```diff
@@ -695,14 +695,15 @@ pub fn parse_file(input_path: &str) -> Result<Arena<String>, ParseError> {
...
-    let yacc_y_path = format!("grammars/{}.y", extension);
+    let mut lex_l_path = format!("grammars/{}.l", extension);
+    let mut yacc_y_path = format!("grammars/{}.y", extension);
     info!("Using lexer: {} and parser: {} for input: {}",
           &lex_l_path,
           &yacc_y_path,
           input_path);
     if !Path::new(&lex_l_path).exists() || !Path::new(&yacc_y_path).exists() {
-        return Err(ParseError::FileNotFound);
+        lex_l_path = String::from("grammars/txt.l");
```
Comment (ltratt) [inline src/lib/ast.rs]: This is a relative path, so doesn't that mean that it won't work if you run the command outside the directory where `grammars` is?

Comment (snim2) [inline src/lib/ast.rs]: You are right, it won't work outside the crate root :(

Comment (ltratt) [inline src/lib/ast.rs]: I suspect that `std::fs::canonicalize` will be part (but not all) of the solution.

**Comment by snim2** [comment]
This isn't quite what I was expecting, but it seems to work.

**Comment by snim2** [comment]
If I move the binary to `$HOME`, I get this:

```sh
$pwd
/home/snim2
snim2)15:37:57(0)~$ ./rstreediff --debug DEBUG revelation_notes.md revelation_notes.md 
INFO:rstreediff: Selecting the GumTree matching algorithm.
Cannot parse .md files. Falling back to plain text parser.
Cannot parse .md files. Falling back to plain text parser.
INFO:treediff::ast: Using lexer: /home/snim2/Desktop/working/softdev/rstreediff/grammars/txt.l and parser: /home/snim2/Desktop/working/softdev/rstreediff/grammars/txt.y for input: revelation_notes.md
INFO:treediff::ast: Using lexer: /home/snim2/Desktop/working/softdev/rstreediff/grammars/txt.l and parser: /home/snim2/Desktop/working/softdev/rstreediff/grammars/txt.y for input: revelation_notes.md
DEBUG:treediff::matchers: find_pos(1)
DEBUG:treediff::matchers: Edit script: INS paragraph  Parent: document 
...
```

So it works fine. If I move the `rstreediff` directory then that fails. This SO answer seems to suggest that there are better solutions to this problem: https://stackoverflow.com/questions/32748918/is-there-any-way-to-include-binary-or-text-files-in-a-rust-library

**Comment by ltratt** [comment]
In all likelihood it only works because your build directory is still in the same place as when you built it. Put another way: embedding compile-time paths into a binary is a guaranteed way to not have a running system on someone else's computer ;)

Including things in the binary is fine, or just look up something like "how do I know what directory my binary was run from?" The solutions are icky, but they're worth knowing about.

**Comment by snim2** [comment]
Since I put the `grammars/*` file in Cargo.toml I had thought it would packages the files up with the binary, so I'm very confused about this. SO seems to think that this macro can be used to read in files at compile time: https://doc.rust-lang.org/std/macro.include_str.html wdyt?

**Comment by ltratt** [comment]
I'd suggest biting the bullet and working out how to find out what directory the binary is being run in. Slurping things into the binary isn't a scalable solution for big files.

**Comment by snim2** [comment]
Just to be pedantically clear, if I am in directory `/home/flibble/` and the binary is in directory `/home/meep/rstreediff/target/debug/` we should expect the grammars to be stored in `/home/meep/rstreediff/grammars`?

**Comment by ltratt** [comment]
Yes.

**Comment by snim2** [comment]
Ok, I *think* this is what you mean...

**Comment by ltratt** [comment]
Yes, this looks like the right sort of thing. I'm not totally certain if we need all the canonicalisation calls though, but admittedly, they won't hurt too much (though they are quite slow IIRC).

**Comment by snim2** [comment]
Ready to squash?

**Comment by ltratt** [comment]
Yep, please squash.

**Comment by snim2** [comment]
Squashed!

**Comment by snim2** [comment]
Thanks

**Pull request merged**